### DISCUSSION TITLE
JSON SCADA MQTT Client

### DISCUSSION THREAD

**Discussion by barsicepy** [riclolsen/json-scada#44] [Q&A] [answered]
Hi Ricardo.

I can't to connect native MQTT Client to the local [mosquito](https://mosquitto.org/man/mosquitto-8.html) broker.
Broker doesn't see any connection from [MQTT Sparkplug-B Driver](https://github.com/riclolsen/json-scada/blob/master/src/mqtt-sparkplug/README.md) client.
But I've connected to it over [Telegraf Driver](https://github.com/riclolsen/json-scada/blob/master/src/telegraf-listener/README.md) very simple.
Could you help me with native driver? [This is settings of connection](https://drive.google.com/file/d/1v6vJVkBfz1WK-Qv9zLwSHlNqeJgLbWrO/view?usp=sharing)

**Comment by riclolsen** [comment, marked as answer]
Hi,

You need to configure a driver instance and run the driver process with a parameter if the instance is not 1. If you need just one instance, keep it as 1. Driver instances are counted separately for each driver type. So you can have instance 1 for MQTT and instance 1 for IEC104 at the same time.

Use this in the command prompt to see logs (stop MQTT driver service first).

    cd \json-scada\src\mqtt-sparkplug
    c:\json-scada\platform-windows\nodejs-runtime\node.exe index.js

If your instance is 2, run with a parameter

    c:\json-scada\platform-windows\nodejs-runtime\node.exe index.js 2

When it works you need to recreate the Windows service accordingly (with the parameter if needed).

Best regards,

Ricardo Olsen

**Reply by barsicepy** [reply]
Yes. 2 instance works thank you. But it can't autocreate tags with values. It create only one tag with TopicName.
JSONPath doesn't help me. What's string I need for get values of JSON string?
This is example of JSON-string, which publishes on the broker with my TopicName:
`{"ActivePower":20.47765,"ReactivePower":12.163717,"PowerFactor":0.8618866,"Frequency":50.18043,"VoltageLine":390.2411,"VoltagePhase":225.6683,"Current":35.47994,"ActiveEnergy":868352.0,"ReactiveEnergy":160743.6}`

P.S. I create 7 topics in driver instance with strings: topicname/$.property and got values, finally. But it is not good method, if we will have 30 variables and over.

P.P.S. How does Windows make to autostart JSON SCADA mqttsparkplugclient service with 1 and 2 instances? Why default service doesn't connect to MQTT client? I could establish connection with `c:\json-scada\platform-windows\nodejs-runtime\node.exe index.js 2` only.

**Reply by riclolsen** [reply]
You can use "topicsScripted" method to subscribe and use scripts to extract payload data.

    https://github.com/riclolsen/json-scada/tree/master/src/mqtt-sparkplug#example-of-script-to-extract-data-from-complex-payloads

The default MQTT service is equivalent to 
    
    cd \json-scada\src\mqtt-sparkplug
    c:\json-scada\platform-windows\nodejs-runtime\node.exe index.js

or 

    c:\json-scada\platform-windows\nodejs-runtime\node.exe index.js 1

Run this to view error messages.

See here examples of creating new services

    https://github.com/riclolsen/json-scada/blob/master/platform-windows/create_services.bat

**Reply by barsicepy** [reply]
> You can use "topicsScripted" method to subscribe and use scripts to extract payload data.

I thought "topic" can extract properties from topic string by [JSONPath](https://github.com/JSONPath-Plus/JSONPath) wildcards. For example string `MyTopicName/$.*~` must extract all property names from root topic. But it doesn't work unfortunately.

**Reply by riclolsen** [reply]
No. But It is a nice idea for future implementation.
